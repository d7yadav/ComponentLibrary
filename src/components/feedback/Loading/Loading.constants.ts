import type { LoadingSpinnerType, LoadingSize, LoadingColor, LoadingVariant } from './Loading.types';

export const LOADING_SPINNER_TYPES: Record<LoadingSpinnerType, LoadingSpinnerType> = {
  circular: 'circular',
  dots: 'dots',
  bars: 'bars',
  pulse: 'pulse',
  bounce: 'bounce',
  ring: 'ring',
  wave: 'wave',
  ripple: 'ripple',
  skeleton: 'skeleton',
} as const;

export const LOADING_SIZES: Record<LoadingSize, LoadingSize> = {
  small: 'small',
  medium: 'medium',
  large: 'large',
} as const;

export const LOADING_COLORS: Record<LoadingColor, LoadingColor> = {
  primary: 'primary',
  secondary: 'secondary',
  error: 'error',
  warning: 'warning',
  info: 'info',
  success: 'success',
  inherit: 'inherit',
} as const;

export const LOADING_VARIANTS: Record<LoadingVariant, LoadingVariant> = {
  default: 'default',
  overlay: 'overlay',
  inline: 'inline',
  button: 'button',
  page: 'page',
} as const;

export const LOADING_SIZE_CONFIGS = {
  small: {
    circular: 20,
    dots: 16,
    bars: 20,
    pulse: 24,
    bounce: 16,
    ring: 24,
    wave: 20,
    ripple: 32,
    fontSize: '0.75rem',
  },
  medium: {
    circular: 32,
    dots: 24,
    bars: 32,
    pulse: 40,
    bounce: 24,
    ring: 40,
    wave: 32,
    ripple: 48,
    fontSize: '0.875rem',
  },
  large: {
    circular: 48,
    dots: 36,
    bars: 48,
    pulse: 60,
    bounce: 36,
    ring: 60,
    wave: 48,
    ripple: 72,
    fontSize: '1rem',
  },
} as const;

export const LOADING_ANIMATION_DURATIONS = {
  circular: 1400,
  dots: 1200,
  bars: 1000,
  pulse: 1500,
  bounce: 600,
  ring: 1200,
  wave: 1600,
  ripple: 1200,
  skeleton: 2000,
} as const;

export const LOADING_ANIMATION_EASINGS = {
  linear: 'linear',
  easeInOut: 'cubic-bezier(0.4, 0, 0.2, 1)',
  bounce: 'cubic-bezier(0.68, -0.55, 0.265, 1.55)',
  elastic: 'cubic-bezier(0.175, 0.885, 0.32, 1.275)',
} as const;

export const LOADING_Z_INDEX = {
  overlay: 1300,
  backdrop: 1200,
  inline: 1,
} as const;

export const LOADING_BACKDROP_OPACITY = {
  light: 0.3,
  medium: 0.5,
  dark: 0.7,
  opaque: 1,
} as const;

export const LOADING_DELAYS = {
  instant: 0,
  short: 200,
  medium: 500,
  long: 1000,
} as const;

export const LOADING_TIMEOUTS = {
  short: 5000,
  medium: 10000,
  long: 30000,
  none: 0,
} as const;

export const ACCESSIBILITY_CONSTANTS = {
  statusRole: 'status',
  progressRole: 'progressbar',
  alertRole: 'alert',
  regionRole: 'region',
  loadingLabel: 'Loading',
  loadingDescription: 'Content is loading, please wait',
  skeletonLabel: 'Loading placeholder',
  liveRegionPolite: 'polite',
  liveRegionAssertive: 'assertive',
} as const;

export const LOADING_DOT_COUNTS = {
  dots: 3,
  bars: 4,
  wave: 5,
  bounce: 3,
} as const;

export const LOADING_MESSAGES = {
  default: 'Loading...',
  data: 'Loading data...',
  page: 'Loading page...',
  content: 'Loading content...',
  image: 'Loading image...',
  video: 'Loading video...',
  form: 'Submitting...',
  save: 'Saving...',
  upload: 'Uploading...',
  download: 'Downloading...',
  processing: 'Processing...',
  validating: 'Validating...',
  authenticating: 'Authenticating...',
} as const;

export const LOADING_PATTERNS = {
  // Page loading
  page: {
    type: LOADING_SPINNER_TYPES.circular,
    variant: LOADING_VARIANTS.overlay,
    size: LOADING_SIZES.large,
    backdrop: true,
    centered: true,
    message: LOADING_MESSAGES.page,
  },
  
  // Content loading
  content: {
    type: LOADING_SPINNER_TYPES.skeleton,
    variant: LOADING_VARIANTS.inline,
    size: LOADING_SIZES.medium,
    fullWidth: true,
  },
  
  // Button loading
  button: {
    type: LOADING_SPINNER_TYPES.circular,
    variant: LOADING_VARIANTS.button,
    size: LOADING_SIZES.small,
    color: LOADING_COLORS.inherit,
  },
  
  // Form submission
  form: {
    type: LOADING_SPINNER_TYPES.dots,
    variant: LOADING_VARIANTS.inline,
    size: LOADING_SIZES.medium,
    message: LOADING_MESSAGES.form,
  },
  
  // Data fetching
  data: {
    type: LOADING_SPINNER_TYPES.pulse,
    variant: LOADING_VARIANTS.default,
    size: LOADING_SIZES.medium,
    centered: true,
    message: LOADING_MESSAGES.data,
  },
  
  // File upload
  upload: {
    type: LOADING_SPINNER_TYPES.bars,
    variant: LOADING_VARIANTS.inline,
    size: LOADING_SIZES.medium,
    message: LOADING_MESSAGES.upload,
  },
  
  // Image loading
  image: {
    type: LOADING_SPINNER_TYPES.ripple,
    variant: LOADING_VARIANTS.overlay,
    size: LOADING_SIZES.medium,
    backdrop: false,
    centered: true,
  },
  
  // Minimal loading
  minimal: {
    type: LOADING_SPINNER_TYPES.dots,
    variant: LOADING_VARIANTS.inline,
    size: LOADING_SIZES.small,
    color: LOADING_COLORS.inherit,
  },
} as const;

export const SKELETON_VARIANTS = {
  text: 'text',
  rectangular: 'rectangular',
  circular: 'circular',
} as const;

export const SKELETON_ANIMATIONS = {
  pulse: 'pulse',
  wave: 'wave',
  none: false,
} as const;

export const SKELETON_DEFAULTS = {
  textHeight: '1.2em',
  textWidth: '100%',
  rectangularHeight: 140,
  rectangularWidth: '100%',
  circularSize: 40,
  lines: 3,
  lineSpacing: 8,
} as const;

export const DEFAULT_LOADING_PROPS = {
  type: LOADING_SPINNER_TYPES.circular,
  size: LOADING_SIZES.medium,
  color: LOADING_COLORS.primary,
  variant: LOADING_VARIANTS.default,
  loading: true,
  backdrop: false,
  backdropOpacity: LOADING_BACKDROP_OPACITY.medium,
  disableBackdropClick: false,
  speed: 1,
  centered: false,
  fullWidth: false,
  fullHeight: false,
  delay: LOADING_DELAYS.instant,
  timeout: LOADING_TIMEOUTS.none,
  zIndex: LOADING_Z_INDEX.overlay,
} as const;