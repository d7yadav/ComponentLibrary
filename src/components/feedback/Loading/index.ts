/**
 * @fileoverview Loading component barrel export
 * @author Dilip Yadav <dilip.sm.yadav@gmail.com>
 */

export { Loading } from './Loading';
export { Skeleton } from './Skeleton';
export type { 
  LoadingProps,
  SkeletonProps,
  LoadingSpinnerType,
  LoadingSize,
  LoadingColor,
  LoadingVariant,
  LoadingStyleProps,
  LoadingSizeConfig,
  LoadingAnimation
} from './Loading.types';
export {
  LOADING_SPINNER_TYPES,
  LOADING_SIZES,
  LOADING_COLORS,
  LOADING_VARIANTS,
  LOADING_SIZE_CONFIGS,
  LOADING_ANIMATION_DURATIONS,
  LOADING_ANIMATION_EASINGS,
  LOADING_Z_INDEX,
  LOADING_BACKDROP_OPACITY,
  LOADING_DELAYS,
  LOADING_TIMEOUTS,
  ACCESSIBILITY_CONSTANTS,
  LOADING_DOT_COUNTS,
  LOADING_MESSAGES,
  LOADING_PATTERNS,
  SKELETON_VARIANTS,
  SKELETON_ANIMATIONS,
  SKELETON_DEFAULTS,
  DEFAULT_LOADING_PROPS,
} from './Loading.constants';
export {
  LoadingContainer,
  LoadingBackdrop,
  LoadingMessage,
  CircularSpinner,
  DotsSpinner,
  BarsSpinner,
  PulseSpinner,
  RingSpinner,
  WaveSpinner,
  RippleSpinner,
  SkeletonText,
  SkeletonRectangular,
  SkeletonCircular,
} from './Loading.styles';