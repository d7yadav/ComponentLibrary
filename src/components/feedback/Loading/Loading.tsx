import { forwardRef, useState, useEffect, memo } from 'react';
import { LoadingProps } from './Loading.types';
import { 
  LoadingContainer,
  LoadingBackdrop,
  LoadingMessage,
  CircularSpinner,
  DotsSpinner,
  BarsSpinner,
  PulseSpinner,
  RingSpinner,
  WaveSpinner,
  RippleSpinner,
} from './Loading.styles';
import { Skeleton } from './Skeleton';
import {
  DEFAULT_LOADING_PROPS,
  ACCESSIBILITY_CONSTANTS,
  LOADING_SIZE_CONFIGS,
  LOADING_DOT_COUNTS,
} from './Loading.constants';

/**
 * Enhanced Loading component with different spinner types and comprehensive options
 * 
 * Features:
 * - 9 spinner types: circular, dots, bars, pulse, bounce, ring, wave, ripple, skeleton
 * - Multiple sizes: small, medium, large with custom size support
 * - Color themes: primary, secondary, error, warning, info, success, inherit
 * - Variants: default, overlay, inline, button, page
 * - Backdrop support with opacity control
 * - Delay and timeout functionality
 * - Accessibility compliance (WCAG 2.1 AA)
 * - TypeScript support with strict typing
 * - Dark theme support via enhanced theme system
 */
export const Loading = forwardRef<HTMLDivElement, LoadingProps>((
  {
    type = DEFAULT_LOADING_PROPS.type,
    size = DEFAULT_LOADING_PROPS.size,
    color = DEFAULT_LOADING_PROPS.color,
    variant = DEFAULT_LOADING_PROPS.variant,
    customSize,
    loading = DEFAULT_LOADING_PROPS.loading,
    message,
    backdrop = DEFAULT_LOADING_PROPS.backdrop,
    backdropOpacity = DEFAULT_LOADING_PROPS.backdropOpacity,
    disableBackdropClick = DEFAULT_LOADING_PROPS.disableBackdropClick,
    speed = DEFAULT_LOADING_PROPS.speed,
    centered = DEFAULT_LOADING_PROPS.centered,
    fullWidth = DEFAULT_LOADING_PROPS.fullWidth,
    fullHeight = DEFAULT_LOADING_PROPS.fullHeight,
    minHeight,
    customColor,
    children,
    delay = DEFAULT_LOADING_PROPS.delay,
    timeout = DEFAULT_LOADING_PROPS.timeout,
    onTimeout,
    zIndex = DEFAULT_LOADING_PROPS.zIndex,
    className,
    sx,
    'aria-label': ariaLabel,
    'aria-describedby': ariaDescribedBy,
    role,
    ...other
  },
  ref
) => {
  const [visible, setVisible] = useState(delay === 0);
  const [timedOut, setTimedOut] = useState(false);

  // Handle delay
  useEffect(() => {
    if (loading && delay > 0) {
      const delayTimer = setTimeout(() => {
        setVisible(true);
      }, delay);

      return () => clearTimeout(delayTimer);
    } else if (!loading) {
      setVisible(false);
    }
  }, [loading, delay]);

  // Handle timeout
  useEffect(() => {
    if (loading && timeout > 0) {
      const timeoutTimer = setTimeout(() => {
        setTimedOut(true);
        onTimeout?.();
      }, timeout);

      return () => clearTimeout(timeoutTimer);
    }
  }, [loading, timeout, onTimeout]);

  // Reset timeout when loading changes
  useEffect(() => {
    if (!loading) {
      setTimedOut(false);
    }
  }, [loading]);

  // Get spinner size
  const getSpinnerSize = () => {
    if (customSize) return customSize;
    const sizeConfig = LOADING_SIZE_CONFIGS[size];
    return sizeConfig[type] || sizeConfig.circular;
  };

  // Render spinner based on type
  const renderSpinner = () => {
    const spinnerSize = getSpinnerSize();
    const spinnerProps = {
      customSize: spinnerSize,
      customColor: customColor,
      customSpeed: speed,
    };

    switch (type) {
      case 'circular':
        return <CircularSpinner {...spinnerProps} />;
      
      case 'dots':
        return (
          <DotsSpinner data-testid="loading" {...spinnerProps}>
            {Array.from({ length: LOADING_DOT_COUNTS.dots }, (_, i) => (
              <div key={i} className="dot" />
            ))}
          </DotsSpinner>
        );
      
      case 'bars':
        return (
          <BarsSpinner {...spinnerProps}>
            {Array.from({ length: LOADING_DOT_COUNTS.bars }, (_, i) => (
              <div key={i} className="bar" />
            ))}
          </BarsSpinner>
        );
      
      case 'pulse':
        return <PulseSpinner {...spinnerProps} />;
      
      case 'bounce':
        return (
          <DotsSpinner data-testid="loading" {...spinnerProps}>
            {Array.from({ length: LOADING_DOT_COUNTS.bounce }, (_, i) => (
              <div key={i} className="dot" />
            ))}
          </DotsSpinner>
        );
      
      case 'ring':
        return <RingSpinner {...spinnerProps} />;
      
      case 'wave':
        return (
          <WaveSpinner {...spinnerProps}>
            {Array.from({ length: LOADING_DOT_COUNTS.wave }, (_, i) => (
              <div key={i} className="wave-bar" />
            ))}
          </WaveSpinner>
        );
      
      case 'ripple':
        return (
          <RippleSpinner {...spinnerProps}>
            <div className="ripple" />
            <div className="ripple" />
          </RippleSpinner>
        );
      
      case 'skeleton':
        return (
          <Skeleton
            variant="rectangular"
            width={fullWidth ? '100%' : spinnerSize * 2}
            height={spinnerSize}
            animation="wave"
          />
        );
      
      default:
        return <CircularSpinner {...spinnerProps} />;
    }
  };

  // Accessibility props
  const accessibilityProps = {
    'aria-label': ariaLabel || ACCESSIBILITY_CONSTANTS.loadingLabel,
    'aria-describedby': ariaDescribedBy,
    role: role || ACCESSIBILITY_CONSTANTS.statusRole,
    'aria-live': ACCESSIBILITY_CONSTANTS.liveRegionPolite,
    'aria-busy': loading,
  };

  // Don't render if not loading and no children, or if delayed
  if ((!loading && !children) || (loading && !visible)) {
    return children ? <>{children}</> : null;
  }

  // Render children when not loading
  if (!loading && children) {
    return <>{children}</>;
  }

  const loadingContent = (
    <LoadingContainer
      ref={ref}
      className={className}
      sx={sx}
      customVariant={variant}
      customCentered={centered}
      customFullWidth={fullWidth}
      customFullHeight={fullHeight}
      customMinHeight={minHeight}
      customZIndex={zIndex}
      {...accessibilityProps}
      {...other}
    >
      {renderSpinner()}
      {message && (
        <LoadingMessage customSize={size}>
          {timedOut ? 'Loading is taking longer than expected...' : message}
        </LoadingMessage>
      )}
    </LoadingContainer>
  );

  // Render with backdrop if specified
  if (backdrop && (variant === 'overlay' || variant === 'page')) {
    return (
      <LoadingBackdrop
        open={loading}
        onClick={disableBackdropClick ? undefined : () => {}}
        customOpacity={backdropOpacity}
        customZIndex={zIndex - 1}
      >
        {loadingContent}
      </LoadingBackdrop>
    );
  }

  // Render with children overlay
  if (children && variant === 'overlay') {
    return (
      <div style={{ position: 'relative' }}>
        {children}
        {loading && loadingContent}
      </div>
    );
  }

  return loadingContent;
});

Loading.displayName = 'Loading';