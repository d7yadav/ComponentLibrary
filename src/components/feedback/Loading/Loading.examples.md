# ðŸ“– Loading Examples

## Basic Usage

```typescript
import { Loading } from '@/components/feedback/Loading';

// Basic usage
const Example = () => {
  return (
    <Loading>
      Basic Loading example
    </Loading>
  );
};
```

## Advanced Examples

```typescript
// Advanced configuration
const AdvancedExample = () => {
  const [value, setValue] = useState('');
  
  return (
    <Loading
      variant="rectangular"
      disabled={false}
      onChange={(e) => setValue(e.target.value)}
      sx={{
        borderRadius: 2,
        '&:hover': {
          transform: 'translateY(-2px)',
        },
      }}
    >
      Advanced Loading
    </Loading>
  );
};
```

## Integration Examples

```typescript
// Integration with form libraries
import { useController, Control } from 'react-hook-form';

interface FormIntegrationProps {
  name: string;
  control: Control;
}

const FormIntegration = ({ name, control }: FormIntegrationProps) => {
  const { field, fieldState } = useController({
    name,
    control,
    defaultValue: '',
  });

  return (
    <Loading
      {...field}
      error={!!fieldState.error}
      helperText={fieldState.error?.message}
    />
  );
};
```

## Edge Cases

```typescript
// Edge case: Empty states
const EmptyState = () => (
  <Loading disabled>
    No content available
  </Loading>
);

// Edge case: Error states
const ErrorState = () => (
  <Loading 
    error
    helperText="Something went wrong"
  />
);

// Edge case: Loading states
const LoadingState = () => (
  <Loading 
    loading
    disabled
  />
);
```

## Performance Optimization Examples

```typescript
// Memoized component for performance
const OptimizedLoading = React.memo((props: LoadingProps) => {
  const memoizedValue = useMemo(() => {
    // Expensive calculation
    return heavyComputation(props.value);
  }, [props.value]);

  return (
    <Loading 
      {...props}
      value={memoizedValue}
    />
  );
});

// Lazy loading for large lists
const LazyList = () => {
  return (
    <VirtualizedList
      itemCount={1000}
      renderItem={({ index }) => (
        <Loading key={index}>
          Item {index}
        </Loading>
      )}
    />
  );
};
```

## Accessibility Examples

```typescript
// Full accessibility implementation
const AccessibleExample = () => {
  return (
    <Loading
      aria-label="Descriptive label for screen readers"
      aria-describedby="helper-text"
      role="button"
      tabIndex={0}
      onKeyDown={(e) => {
        if (e.key === 'Enter' || e.key === ' ') {
          handleAction();
        }
      }}
    >
      Accessible Loading
    </Loading>
  );
};

// High contrast mode support
const HighContrastExample = () => {
  return (
    <Loading
      sx={{
        '@media (prefers-contrast: high)': {
          borderWidth: 2,
          borderColor: 'text.primary',
        },
      }}
    >
      High contrast support
    </Loading>
  );
};
```

---

*These examples demonstrate real-world usage patterns and best practices for the Loading component.*
