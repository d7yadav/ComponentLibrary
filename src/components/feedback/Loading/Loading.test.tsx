/**
 * @fileoverview Loading Component Tests
 * @author AI Generated Tests - 2025-07-13
 */

import { ThemeProvider , createTheme } from '@mui/material/styles';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { axe, toHaveNoViolations } from 'jest-axe';

import { Loading } from './Loading';
import { LOADING_CONSTANTS } from './Loading.constants';
import type { LoadingProps } from './Loading.types';

// Add jest-axe matcher
expect.extend(toHaveNoViolations);

// Test theme
const testTheme = createTheme();

// Test wrapper component
const TestWrapper = ({ children }: { children: React.ReactNode }) => (
  <ThemeProvider theme={testTheme}>{children}</ThemeProvider>
);

// Helper function to render component with theme
const renderWithTheme = (ui: React.ReactElement, options = {}) => {
  return render(ui, { wrapper: TestWrapper, ...options });
};

describe('Loading Component', () => {
  // Basic rendering tests
  describe('Rendering', () => {
    it('renders without crashing', () => {
      renderWithTheme(<Loading>Test content</Loading>);
      expect(screen.getByRole('status')).toBeInTheDocument();
          });

    it('renders with message correctly', () => {
      const testMessage = 'Loading data...';
      renderWithTheme(<Loading message={testMessage} />);
      expect(screen.getByText(testMessage)).toBeInTheDocument();
    });

    it('renders with custom test id', () => {
      const testId = 'custom-loading';
      renderWithTheme(
        <Loading data-testid={testId}>Test</Loading>
      );
      expect(screen.getByTestId(testId)).toBeInTheDocument();
    });

    it('applies custom className', () => {
      const customClass = 'custom-class';
      renderWithTheme(
        <Loading className={customClass}>Test</Loading>
      );
      expect(screen.getByRole('status')).toHaveClass(customClass);
    });
  });

  // Props testing
  describe('Props', () => {
    describe('variant prop', () => {
      const variants: LoadingProps['variant'][] = [
        'circular',
        'dots',
        'bars',
        'pulse',
        'bounce',
        'ring',
        'wave',
        'ripple',
        'skeleton',
      ];

      variants.forEach((variant) => {
        it(`renders ${variant} variant correctly`, () => {
          renderWithTheme(
            <Loading variant={variant} data-testid="variant-test">
              {variant} variant
            </Loading>
          );
          const element = screen.getByTestId('variant-test');
          expect(element).toBeInTheDocument();
        });
      });
    });

    describe('size prop', () => {
      const sizes: LoadingProps['size'][] = [
        'small',
        'medium',
        'large',
      ];

      sizes.forEach((size) => {
        it(`renders ${size} size correctly`, () => {
          renderWithTheme(
            <Loading size={size} data-testid="size-test">
              {size} size
            </Loading>
          );
          const element = screen.getByTestId('size-test');
          expect(element).toBeInTheDocument();
        });
      });
    });



    it('handles loading active state correctly', () => {
      renderWithTheme(
        <Loading loading={true} data-testid="active-test" />
      );
      
      const element = screen.getByTestId('active-test');
      expect(element).toBeInTheDocument();
      expect(screen.getByRole('status')).toBeInTheDocument();
    });
  });

  // Interaction testing
  describe('Interactions', () => {
    it('handles loading state changes', async () => {
      const { rerender } = renderWithTheme(
        <Loading loading={false} data-testid="loading-component">
          Content
        </Loading>
      );
      
      // Should show content when not loading
      expect(screen.getByText('Content')).toBeInTheDocument();
      expect(screen.queryByRole('status')).not.toBeInTheDocument();
      
      // Switch to loading state
      rerender(
        <TestWrapper>
          <Loading loading={true} data-testid="loading-component">
            Content
          </Loading>
        </TestWrapper>
      );
      
      // Should show spinner when loading
      await waitFor(() => {
        expect(screen.getByRole('status')).toBeInTheDocument();
      });
    });

    it('handles delayed loading appearance', async () => {
      renderWithTheme(
        <Loading 
          loading={true} 
          delay={100} 
          data-testid="delayed-loading"
        >
          Content
        </Loading>
      );
      
      // Initially should not show spinner
      expect(screen.queryByRole('progressbar')).not.toBeInTheDocument();
      
      // Should appear after delay
      await waitFor(() => {
        expect(screen.getByRole('status')).toBeInTheDocument();
      }, { timeout: 200 });
    });

    it('handles timeout functionality', async () => {
      const handleTimeout = vi.fn();
      
      renderWithTheme(
        <Loading 
          loading={true} 
          timeout={100} 
          onTimeout={handleTimeout}
          data-testid="timeout-loading"
        >
          Content
        </Loading>
      );
      
      // Should call timeout handler after timeout period
      await waitFor(() => {
        expect(handleTimeout).toHaveBeenCalledTimes(1);
      }, { timeout: 200 });
    });

    it('handles backdrop clicks when enabled', async () => {
      const user = userEvent.setup();
      const handleBackdropClick = vi.fn();
      
      renderWithTheme(
        <Loading 
          loading={true}
          variant="overlay"
          backdrop={true}
          disableBackdropClick={false}
          onClick={handleBackdropClick}
          data-testid="backdrop-loading"
        >
          Content
        </Loading>
      );
      
      const backdrop = screen.getByTestId('backdrop-loading');
      await user.click(backdrop);
      
      expect(handleBackdropClick).toHaveBeenCalledTimes(1);
    });
  });



  // Accessibility testing
  describe('Accessibility', () => {
    it('meets WCAG accessibility guidelines', async () => {
      const { container } = renderWithTheme(
        <Loading>Accessible Loading</Loading>
      );
      
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });

    it('supports screen readers', () => {
      const ariaLabel = 'Screen reader label';
      renderWithTheme(
        <Loading aria-label={ariaLabel}>
          Loading for screen readers
        </Loading>
      );
      
      expect(screen.getByLabelText(ariaLabel)).toBeInTheDocument();
    });

    it('supports keyboard navigation', () => {
      renderWithTheme(<Loading tabIndex={0}>Keyboard Loading</Loading>);
      
      const element = screen.getByRole('status');
      element.focus();
      expect(element).toHaveFocus();
    });

    it('provides proper ARIA attributes', () => {
      renderWithTheme(
        <Loading 
          aria-describedby="description"
          aria-controls="controlled-element"
        >
          ARIA Loading
        </Loading>
      );
      
      const element = screen.getByRole('status');
      expect(element).toHaveAttribute('aria-describedby', 'description');
      expect(element).toHaveAttribute('aria-controls', 'controlled-element');
    });
  });

  // Theme integration testing
  describe('Theme Integration', () => {
    it('respects theme customizations', () => {
      const customTheme = createTheme({
        components: {
          MuiLoading: {
            defaultProps: {
              variant: 'circular',
            },
            styleOverrides: {
              root: {
                backgroundColor: 'rgb(255, 0, 0)',
              },
            },
          },
        },
      });

      render(
        <ThemeProvider theme={customTheme}>
          <Loading data-testid="themed-loading">
            Themed Loading
          </Loading>
        </ThemeProvider>
      );

      const element = screen.getByTestId('themed-loading');
      expect(element).toBeInTheDocument();
    });

    it('supports dark mode', () => {
      const darkTheme = createTheme({
        palette: {
          mode: 'dark',
        },
      });

      render(
        <ThemeProvider theme={darkTheme}>
          <Loading>Dark Mode Loading</Loading>
        </ThemeProvider>
      );

      expect(screen.getByRole('status')).toBeInTheDocument();
    });
  });

  // Performance testing
  describe('Performance', () => {
    it('renders efficiently with many instances', () => {
      const manyComponents = Array.from({ length: 100 }, (_, i) => (
        <Loading key={i}>Loading {i}</Loading>
      ));

      const { container } = renderWithTheme(<div>{manyComponents}</div>);
      expect(container.querySelectorAll('[role="status"]')).toHaveLength(100);
    });

    it('respects reduced motion preferences', () => {
      // Mock prefers-reduced-motion
      Object.defineProperty(window, 'matchMedia', {
        writable: true,
        value: vi.fn().mockImplementation((query) => ({
          matches: query === '(prefers-reduced-motion: reduce)',
          media: query,
          onchange: null,
          addListener: vi.fn(),
          removeListener: vi.fn(),
          addEventListener: vi.fn(),
          removeEventListener: vi.fn(),
          dispatchEvent: vi.fn(),
        })),
      });

      renderWithTheme(
        <Loading animate={false}>
          No Animation Loading
        </Loading>
      );

      expect(screen.getByRole('status')).toBeInTheDocument();
    });
  });

  // Edge cases and error handling
  describe('Edge Cases', () => {
    it('handles empty children gracefully', () => {
      renderWithTheme(<Loading />);
      expect(screen.getByRole('status')).toBeInTheDocument();
    });

    it('handles null children gracefully', () => {
      renderWithTheme(<Loading>{null}</Loading>);
      expect(screen.getByRole('status')).toBeInTheDocument();
    });

    it('handles undefined props gracefully', () => {
      renderWithTheme(
        <Loading 
          variant={undefined as any} 
          size={undefined as any}
        >
          Undefined Props
        </Loading>
      );
      expect(screen.getByRole('status')).toBeInTheDocument();
    });

    it('handles complex configurations', () => {
      renderWithTheme(
        <Loading
          type="circular"
          size="small"
          loading={true}
          speed={1}
        >
          Complex Loading
        </Loading>
      );
      
      expect(screen.getByRole('status')).toBeInTheDocument();
    });
  });

  // Snapshot testing
  describe('Snapshots', () => {
    it('matches snapshot for default props', () => {
      const { container } = renderWithTheme(
        <Loading>Default Loading</Loading>
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches snapshot for all variants', () => {
      const variants: LoadingProps['variant'][] = [
        'circular',
        'dots',
        'bars',
        'pulse',
        'bounce',
        'ring',
        'wave',
        'ripple',
        'skeleton',
      ];

      variants.forEach((variant) => {
        const { container } = renderWithTheme(
          <Loading variant={variant}>{variant} Loading</Loading>
        );
        expect(container.firstChild).toMatchSnapshot(`variant-${variant}`);
      });
    });
  });
});
