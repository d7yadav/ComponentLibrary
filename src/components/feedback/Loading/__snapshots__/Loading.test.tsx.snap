// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Loading Component > Snapshots > matches snapshot for all variants > variant-bars 1`] = `
<div
  aria-busy="true"
  aria-label="Loading"
  aria-live="polite"
  class="MuiBox-root css-fc7vi2"
  role="status"
>
  <div
    class="MuiBox-root css-i0usn6"
  />
</div>
`;

exports[`Loading Component > Snapshots > matches snapshot for all variants > variant-bounce 1`] = `
<div
  aria-busy="true"
  aria-label="Loading"
  aria-live="polite"
  class="MuiBox-root css-fc7vi2"
  role="status"
>
  <div
    class="MuiBox-root css-i0usn6"
  />
</div>
`;

exports[`Loading Component > Snapshots > matches snapshot for all variants > variant-circular 1`] = `
<div
  aria-busy="true"
  aria-label="Loading"
  aria-live="polite"
  class="MuiBox-root css-fc7vi2"
  role="status"
>
  <div
    class="MuiBox-root css-i0usn6"
  />
</div>
`;

exports[`Loading Component > Snapshots > matches snapshot for all variants > variant-dots 1`] = `
<div
  aria-busy="true"
  aria-label="Loading"
  aria-live="polite"
  class="MuiBox-root css-fc7vi2"
  role="status"
>
  <div
    class="MuiBox-root css-i0usn6"
  />
</div>
`;

exports[`Loading Component > Snapshots > matches snapshot for all variants > variant-pulse 1`] = `
<div
  aria-busy="true"
  aria-label="Loading"
  aria-live="polite"
  class="MuiBox-root css-fc7vi2"
  role="status"
>
  <div
    class="MuiBox-root css-i0usn6"
  />
</div>
`;

exports[`Loading Component > Snapshots > matches snapshot for all variants > variant-ring 1`] = `
<div
  aria-busy="true"
  aria-label="Loading"
  aria-live="polite"
  class="MuiBox-root css-fc7vi2"
  role="status"
>
  <div
    class="MuiBox-root css-i0usn6"
  />
</div>
`;

exports[`Loading Component > Snapshots > matches snapshot for all variants > variant-ripple 1`] = `
<div
  aria-busy="true"
  aria-label="Loading"
  aria-live="polite"
  class="MuiBox-root css-fc7vi2"
  role="status"
>
  <div
    class="MuiBox-root css-i0usn6"
  />
</div>
`;

exports[`Loading Component > Snapshots > matches snapshot for all variants > variant-skeleton 1`] = `
<div
  aria-busy="true"
  aria-label="Loading"
  aria-live="polite"
  class="MuiBox-root css-fc7vi2"
  role="status"
>
  <div
    class="MuiBox-root css-i0usn6"
  />
</div>
`;

exports[`Loading Component > Snapshots > matches snapshot for all variants > variant-wave 1`] = `
<div
  aria-busy="true"
  aria-label="Loading"
  aria-live="polite"
  class="MuiBox-root css-fc7vi2"
  role="status"
>
  <div
    class="MuiBox-root css-i0usn6"
  />
</div>
`;

exports[`Loading Component > Snapshots > matches snapshot for default props 1`] = `
<div
  aria-busy="true"
  aria-label="Loading"
  aria-live="polite"
  class="MuiBox-root css-fc7vi2"
  role="status"
>
  <div
    class="MuiBox-root css-i0usn6"
  />
</div>
`;
